### ISSUE TITLE
[BUG] Error with JAX-JIT, finite shot, qml.expval(qml.Hermitian(matrix))

### ISSUE THREAD

**Issue by ankit27kh** [PennyLaneAI/pennylane#2406]
### Expected behavior

This should work:
```
dev = qml.device('default.qubit.jax', wires=2, shots=10)
projector = np.zeros((2 ** 2, 2 ** 2))
projector[0][0] = 1

@jax.jit
@qml.qnode(dev, interface='jax')
def circ():
    qml.PauliZ(wires=0)
    return qml.expval(qml.Hermitian(projector, wires=range(2)))

print(circ())
```

### Actual behavior

Code errors out

### Additional information

Works with `shots=None`

### Tracebacks

```shell
---------------------------------------------------------------------------

UnfilteredStackTrace                      Traceback (most recent call last)

<ipython-input-7-012dd5554f09> in <module>()
     10 
---> 11 print(circ())

49 frames

UnfilteredStackTrace: jax._src.errors.TracerArrayConversionError: The numpy.ndarray conversion method __array__() was called on the JAX Tracer object Traced<ShapedArray(int32[])>with<DynamicJaxprTrace(level=0/2)>
While tracing the function circ at <ipython-input-7-012dd5554f09>:5 for jit, this value became a tracer due to JAX operations on these lines:

  operation a:c64[4] = convert_element_type[new_dtype=complex64 weak_type=False] b
    from line /usr/local/lib/python3.7/dist-packages/pennylane/devices/default_qubit.py:590 (_create_basis_state)

  operation a:bool[] = lt b b
    from line /usr/local/lib/python3.7/dist-packages/pennylane/devices/default_qubit.py:463 (_apply_phase)

  operation a:i32[] = add b c
    from line /usr/local/lib/python3.7/dist-packages/pennylane/devices/default_qubit.py:463 (_apply_phase)

  operation a:i32[] = select_n b c d
    from line /usr/local/lib/python3.7/dist-packages/pennylane/devices/default_qubit.py:463 (_apply_phase)

  operation a:bool[] = lt b c
    from line /usr/local/lib/python3.7/dist-packages/pennylane/devices/default_qubit.py:463 (_apply_phase)

(Additional originating lines are not shown.)
See https://jax.readthedocs.io/en/latest/errors.html#jax.errors.TracerArrayConversionError

The stack trace below excludes JAX-internal frames.
The preceding is the original exception that occurred, unmodified.

--------------------


The above exception was the direct cause of the following exception:

TracerArrayConversionError                Traceback (most recent call last)

/usr/local/lib/python3.7/dist-packages/pennylane/_qubit_device.py in sample(self, observable, shot_range, bin_size)
    844             indices = samples @ powers_of_two
    845             try:
--> 846                 samples = observable.get_eigvals()[indices]
    847             except qml.operation.EigvalsUndefinedError as e:
    848                 # if observable has no info on eigenvalues, we cannot return this measurement

TracerArrayConversionError: The numpy.ndarray conversion method __array__() was called on the JAX Tracer object Traced<ShapedArray(int32[])>with<DynamicJaxprTrace(level=0/2)>
While tracing the function circ at <ipython-input-7-012dd5554f09>:5 for jit, this value became a tracer due to JAX operations on these lines:

  operation a:c64[4] = convert_element_type[new_dtype=complex64 weak_type=False] b
    from line /usr/local/lib/python3.7/dist-packages/pennylane/devices/default_qubit.py:590 (_create_basis_state)

  operation a:bool[] = lt b b
    from line /usr/local/lib/python3.7/dist-packages/pennylane/devices/default_qubit.py:463 (_apply_phase)

  operation a:i32[] = add b c
    from line /usr/local/lib/python3.7/dist-packages/pennylane/devices/default_qubit.py:463 (_apply_phase)

  operation a:i32[] = select_n b c d
    from line /usr/local/lib/python3.7/dist-packages/pennylane/devices/default_qubit.py:463 (_apply_phase)

  operation a:bool[] = lt b c
    from line /usr/local/lib/python3.7/dist-packages/pennylane/devices/default_qubit.py:463 (_apply_phase)

(Additional originating lines are not shown.)
See https://jax.readthedocs.io/en/latest/errors.html#jax.errors.TracerArrayConversionError
```

### System information

```shell
Name: PennyLane
Version: 0.22.1
Summary: PennyLane is a Python quantum machine learning library by Xanadu Inc.
Home-page: https://github.com/XanaduAI/pennylane
Author: None
Author-email: None
License: Apache License 2.0
Location: /usr/local/lib/python3.7/dist-packages
Requires: toml, cachetools, autoray, scipy, appdirs, numpy, pennylane-lightning, retworkx, semantic-version, autograd, networkx
Required-by: PennyLane-Lightning
Platform info:           Linux-5.4.144+-x86_64-with-Ubuntu-18.04-bionic
Python version:          3.7.13
Numpy version:           1.21.5
Scipy version:           1.4.1
Installed devices:
```

### Existing GitHub issues

- [X] I have searched existing GitHub issues to make sure the issue does not already exist.

**Comment by CatalinaAlbornoz** [comment]
Hi @ankit27kh, thank you for reporting this bug. We will take a look into it.

**Comment by antalszava** [comment]
Hi @ankit27kh, until we have a definite fix to the issue raised, specifically for projectors a potential fix in the meantime could be to use `qml.Projector`:
```python
dev = qml.device('default.qubit.jax', wires=2, shots=10)

@jax.jit
@qml.qnode(dev, interface='jax')
def circ():
    qml.PauliZ(wires=0)
    return qml.expval(qml.Projector(np.array([0,0]), wires=range(2)))

print(circ())
```

**Comment by antalszava** [comment]
Hi @ankit27kh, we're adding a fix for the original issue in #2427, thanks for catching it :pray: 

One note: the projector object will have to be passed into the QNode such that that's being jitted too (and we can detect within PennyLane that we should apply the `jax-jit` interface):
```python
import pennylane as qml
import jax

from pennylane import numpy as np

dev = qml.device('default.qubit.jax', wires=2, shots=10)
projector = np.zeros((2 ** 2, 2 ** 2))
projector[0][0] = 1

@jax.jit
@qml.qnode(dev, interface='jax')
def circ(projector):
    qml.PauliZ(wires=0)
    return qml.expval(qml.Hermitian(projector, wires=range(2)))

print(circ(projector))
```
